src/tss2-tcti/tcti-libtpms.c

```diff
@@ -0,0 +1,969 @@
+ * All rights reserved.
+ ******************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include <dlfcn.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/syscall.h>
+#include <netinet/in.h>
+#include <pthread.h>
+#include "tss2_tcti_libtpms.h"
+
+#include "tcti-libtpms.h"
+#include "tcti-common.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * libtpms API calls need to be wrapped. We set the current active TCTI module
+ * for this thread. This is needed because libtpms may call callbacks and these
+ * need to know which TCTI context they have to operate on.
+ *
+ * This macro assumes that TSS2_RC rc and int ret are declared. Jumps to
+ * fail_label on error. In this case, rc contains the respective error code.
+ */
+#define LIBTPMS_API_CALL(fail_label, tcti_libtpms, function, ...) \
+    rc = tcti_libtpms_set_current_tcti(tcti_libtpms); \
+    if (rc != TPM2_RC_SUCCESS)  { \
+        goto fail_label; \
+    } \
+    ret = tcti_libtpms->function(__VA_ARGS__); \
+    if (ret != TPM_SUCCESS) { \
+        LOG_ERROR("libtpms function " #function "() failed with return code 0x%" PRIx32, ret); \
+        rc = TSS2_TCTI_RC_GENERAL_FAILURE; \
+        goto fail_label; \
+    } \
+    rc = tcti_libtpms_clear_current_tcti(); \
+    if (rc != TPM2_RC_SUCCESS)  { \
+        goto fail_label; \
+    }
+
+typedef struct {
+    pid_t thread_id;
+    TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms;
+} tcti_instance;
+
+/*
+ * List which associates each registered thread with the currently active
+ * libtpms tcti.
+ */
+static tcti_instance current_libtpms_tctis[64] = {0};
+pthread_mutex_t current_libtpms_tctis_lock;
+
+/*
+ * Return the index of this thread in the list of thread / current TCTI
+ * associations.
+ *
+ * If neccessary, register thread id in the global list. This list tracks the
+ * currently active TCTI per thread. This is needed to inform the libtpms
+ * callbacks which TCTI context they have to operate on.
+ *
+ * Registering is the only write operation on global data that can happen from
+ * multiple threads at once. Therefore, we have to lock this function with a
+ * pthread.
+ *
+ * TODO: as of now, threads cannot be unregistered. After list fills up, no new
+ * libtpms TCTI instances can be created in unregistered threads.
+ */
+static int
+tcti_libtpms_register_get_thread_index(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
+{
+    pid_t thread_id = syscall(__NR_gettid);
+    int index;
+
+    /* return if thread is already registered */
+    for (size_t i = 0; i < ARRAY_LEN(current_libtpms_tctis); i++) {
+        if (current_libtpms_tctis[i].thread_id == thread_id) {
+            return i;
+        }
+    }
+
+    /* did not find thread in the list, for write access we need to lock */
+    pthread_mutex_lock(&current_libtpms_tctis_lock);
```

Comment:
The part above the mutex (searching for the current thread) is read-only and not problematic.

Everything below the mutex goes through the list anew and does not rely on anything read before (it could have changed since). Therefore I do not think this is racy.

Reply:
you need to handle the lock failing.

Reply:
Thanks, that's it.